tools/ktx/validate.cpp

```diff
@@ -0,0 +1,1577 @@
+    uint32_t model: 8;
+    uint32_t primaries: 8;
+    uint32_t transfer: 8;
+    uint32_t flags: 8;
+    uint32_t texelBlockDimension0: 8;
+    uint32_t texelBlockDimension1: 8;
+    uint32_t texelBlockDimension2: 8;
+    uint32_t texelBlockDimension3: 8;
+    std::array<uint8_t, 8> bytesPlanes;
+
+    [[nodiscard]] bool matchTexelBlockDimensions(uint8_t dim0, uint8_t dim1, uint8_t dim2, uint8_t dim3) const {
+        return texelBlockDimension0 == dim0
+                && texelBlockDimension1 == dim1
+                && texelBlockDimension2 == dim2
+                && texelBlockDimension3 == dim3;
+    }
+
+    [[nodiscard]] bool hasNonZeroBytePlane() const {
+        return bytesPlanes[0] != 0 || bytesPlanes[1] != 0
+                || bytesPlanes[2] != 0 || bytesPlanes[3] != 0
+                || bytesPlanes[4] != 0 || bytesPlanes[5] != 0
+                || bytesPlanes[6] != 0 || bytesPlanes[7] != 0;
+    }
+};
+static_assert(sizeof(BDFD) == 24);
+
+struct SampleType {
+    uint32_t bitOffset: 16;
+    uint32_t bitLength: 8;
+    uint32_t channelType: 4;
+    uint32_t qualifierLinear: 1;
+    uint32_t qualifierExponent: 1;
+    uint32_t qualifierSigned: 1;
+    uint32_t qualifierFloat: 1;
+    uint32_t samplePosition0: 8;
+    uint32_t samplePosition1: 8;
+    uint32_t samplePosition2: 8;
+    uint32_t samplePosition3: 8;
+    uint32_t lower;
+    uint32_t upper;
+};
+static_assert(sizeof(SampleType) == 16);
+
+/// RAII Handler for ktxTexture
+template <typename T>
+class KTXTexture final {
+private:
+    T* _handle;
+
+public:
+    explicit KTXTexture(std::nullptr_t) : _handle{nullptr} { }
+    explicit KTXTexture(T* handle) : _handle{handle} { }
+
+    KTXTexture(const KTXTexture&) = delete;
+    KTXTexture& operator=(const KTXTexture&) = delete;
+
+    KTXTexture(KTXTexture&& other) noexcept : _handle{other._handle} {
+        other._handle = nullptr;
+    }
+
+    KTXTexture& operator=(KTXTexture&& toMove) & {
+        _handle = toMove._handle;
+        toMove._handle = nullptr;
+        return *this;
+    }
+
+    ~KTXTexture() {
+        if (_handle) {
+            ktxTexture_Destroy(handle<ktxTexture>());
+            _handle = nullptr;
+        }
+    }
+
+    template <typename U = T>
+    inline U* handle() const {
+        return reinterpret_cast<U*>(_handle);
+    }
+
+    template <typename U = T>
+    inline U** pHandle() {
+        return reinterpret_cast<U**>(&_handle);
+    }
+
+    inline operator T*() const {
+        return _handle;
+    }
+};
+
+class FatalValidationError : public std::runtime_error {
+public:
+    ValidationReport report;
+
+public:
+    explicit FatalValidationError(ValidationReport report) :
+        std::runtime_error(report.details),
+        report(std::move(report)) {}
+};
+
+static constexpr int RETURN_CODE_VALIDATION_FAILURE = 3;
+static constexpr int RETURN_CODE_VALIDATION_SUCCESS = 0;
+static constexpr int MAX_NUM_KV_ENTRY = 100;
+static constexpr int MAX_NUM_DFD_BLOCK = 10;
```

Comment:
It is still not clear to me how having these limits helps as you can already tell if the dfd and kvd byte lengths are ridiculous. If we continue with them, please ensure any related messages make clear these are validator limits not spec. limits.

Reply:
Limiting KV pairs and DFD Blocks are just fail safes to stop validation spewing error messages if it ran into a corrupt/garbage file.
Note that the original ktx2check had the same error check for KV pairs (with 75 as limit).
Its not required, but a reasonable thing to have.

I will clarify the messages.
Should I change the related errors to warnings? (I was already thinking about this one)

Reply:
Keep the limits.  I agree with changing the related errors to warnings. The language still needs to be clear about the source of the limit that is being violated.